# Example 1 - CHapter 4 

Leather Limited manufactures two types of belts: the deluxe model and the regular model.
Each type requires 1 sq yd of leather. A regular belt requires 1 hour of skilled labor, and
a deluxe belt requires 2 hours. Each week, 40 sq yd of leather and 60 hours of skilled la-
bor are available. Each regular belt contributes \$3 to profit and each deluxe belt, 
\$4. If
we define

$x1$ = number of deluxe belts produced weekly

$x2$ = number of regular belts produced weekly

the appropriate LP is

max z = 4x1 + 3x2

s.t.

x1 = x2 + 40 (Leather constraint)

2x1 = x2 + 60 (Labor constraint)

x1, x2 >= 0

In [17]:
from ortools.init import pywrapinit

pywrapinit.CppBridge.InitLogging('Operations-Research-Winston.ipynb')
cpp_flags = pywrapinit.CppFlags()
cpp_flags.logtostderr = True
cpp_flags.log_prefix = False
pywrapinit.CppBridge.SetFlags(cpp_flags)

: 

: 

In [ ]:
import pyomo.environ as pyo
from pyomo.opt import SolverFactory
import pyomo.core as pyc

model = pyc.ConcreteModel()

model.x1 = pyc.Var(within=pyc.NonNegativeReals)
x1 = model.x1
model.x2 = pyc.Var(within=pyc.NonNegativeReals)
x2 = model.x2

model.Obj = pyc.Objective(expr = 4*x1 +3*x2, sense = pyc.maximize) 

model.Const1 = pyc.Constraint(expr=x1+x2 <= 40)
model.Const2 = pyc.Constraint(expr=2*x1+x2 <= 60)

opm = SolverFactory("glpk")
results = opm.solve(model)

print(f"Objective function = {model.Obj()}")
print(f"x1 = {model.x1()}")
print(f"x2 = {model.x2()}")
print("-----------------")
# print(results)


Objective function = 140.0
x1 = 20.0
x2 = 20.0
-----------------


In [ ]:
from ortools.linear_solver import pywraplp

# Create the linear solver with the GLOP backend.
solver: pywraplp.Solver = pywraplp.Solver.CreateSolver('GLOP') # third party solvers require building ortools from source
if not solver:
    raise

x1: pywraplp.Variable = solver.IntVar(0, solver.infinity(), 'x1')
x2: pywraplp.Variable = solver.IntVar(0, solver.infinity(), 'x2')

print('Number of variables =', solver.NumVariables())

# # Create a linear constraint, 0 <= x + y <= 2.
solver.Add(x1+x2 <= 40)
solver.Add(2*x1+x2 <= 60)

print('Number of constraints =', solver.NumConstraints())

solver.Maximize(4*x1 +3*x2) 

status = solver.Solve()

if status == pywraplp.Solver.OPTIMAL:
    print('Solution:')
    print(f'Objective value = {solver.Objective().Value():.2f}')
    print(f'x1 = {x1.solution_value():.2f}')
    print(f'x2 = {x2.solution_value():.2f}')
    print('\nAdvanced usage:')
    print('Problem solved in %f milliseconds' % solver.wall_time())
    print('Problem solved in %d iterations' % solver.iterations())
else:
    print('The problem does not have an optimal solution.')

RuntimeError: No active exception to reraise

# Example 2 - Page 140

The Dakota Furniture Company manufactures desks, tables, and chairs. The manufacture
of each type of furniture requires lumber and two types of skilled labor: finishing and carpentry. The amount of each resource needed to make each type of furniture is given in
Table 4.
Currently, 48 board feet of lumber, 20 finishing hours, and 8 carpentry hours are avail-
able. A desk sells for \$60, a table for \$30, and a chair for $20. Dakota believes that demand for desks and chairs is unlimited, but at most five tables can be sold. Because the
available resources have already been purchased, Dakota wants to maximize total
revenue.

Resource|Desk|Table|Chair
-|-|-|-
Lumber (board ft)|8|6.5|1.5
Finishing hours|4|2.5|1.5
Carpentry hours|2|1.5|0.5

Defining the decision variables as

- $x1$ = number of desks produced
- $x2$ = number of tables produced
- $x3$ = number of chairs produced

it is easy to see that Dakota should solve the following LP:

max $z = 60x1 + 30x2 + 20x3$

s.t.
- 8x1 + 6x2 + x3 <= 48 (Lumber constraint)
- 4x1 + 2x2 + 1.5x3 <= 20 (Finishing constraint)
- 2x1 + 1.5x2 + 0.5x3 <= 8 (Carpentry constraint)
- x2 <= 5 (Limitation on table demand)
- x1, x2, x3 >= 0